cedar-policy/src/api.rs

```diff
@@ -627,80 +627,82 @@ impl Schema {
+    #[error("cycle in action hierarchy")]
     CycleInActionHierarchy,
     /// Parse errors occurring while parsing an entity type.
-    #[error("Parse error in entity type: {0}")]
-    EntityTypeParse(ParseErrors),
+    #[error("parse error in entity type: {0}")]
+    EntityTypeParseError(ParseErrors),
     /// Parse errors occurring while parsing a namespace identifier.
-    #[error("Parse error in namespace identifier: {0}")]
-    NamespaceParse(ParseErrors),
-    /// Parse errors occurring while parsing a common type identifier.
-    #[error("Parse error in common type identifier: {0}")]
-    CommonTypeParseError(ParseErrors),
+    #[error("parse error in namespace identifier: {0}")]
+    NamespaceParseError(ParseErrors),
     /// Parse errors occurring while parsing an extension type.
-    #[error("Parse error in extension type: {0}")]
-    ExtensionTypeParse(ParseErrors),
+    #[error("parse error in extension type: {0}")]
+    ExtensionTypeParseError(ParseErrors),
+    /// Parse errors occurring while parsing the name of one of reusable
+    /// declared types.
+    #[error("parse error in common type identifier: {0}")]
+    CommonTypeParseError(ParseErrors),
     /// The schema file included an entity type `Action` in the entity type
     /// list. The `Action` entity type is always implicitly declared, and it
     /// cannot currently have attributes or be in any groups, so there is no
     /// purposes in adding an explicit entry.
-    #[error("Entity type `Action` declared in `entityTypes` list.")]
+    #[error("entity type `Action` declared in `entityTypes` list")]
     ActionEntityTypeDeclared,
-    /// One or more action entities are declared with `attributes` lists, but
-    /// action entities cannot have attributes.
-    #[error("Actions declared with `attributes`: [{}]", .0.iter().map(String::as_str).join(", "))]
-    ActionEntityAttributes(Vec<String>),
-    /// An action context or entity type shape was declared to have a type other
-    /// than `Record`.
-    #[error("{0} is not a record")]
+    /// One or more action entities are declared with `attributes`, but this is
+    /// not currently supported.
+    #[error("action declared with `attribute`: [{}]", .0.iter().join(", "))]
+    ActionHasAttributes(Vec<String>),
+    /// `context` or `shape` fields are not records
+    #[error("`{0}` is not a record")]
     ContextOrShapeNotRecord(ContextOrShape),
-    /// An Action Entity (transitively) has an attribute that is an empty set
-    #[error("Action attribute is an empty set")]
-    ActionEntityAttributeEmptySet,
-    /// An Action Entity (transitively) has an attribute of unsupported type (ExprEscape, EntityEscape or ExtnEscape)
+    /// An action entity (transitively) has an attribute that is an empty set.
+    /// This error variant should only be used when `PermitAttributes` is enabled.
+    #[error("action `{0}` has an attribute that is an empty set")]
+    ActionAttributesContainEmptySet(EntityUid),
```

Comment:
It would be helpful if this comment conveyed why this is an error. I assume the problem is that the validator can't assign a type to the empty set?